coreos-installer

```diff
@@ -664,31 +682,87 @@ log() {
     echo "$1" >&2
 }
 
+#########################################################
+#Update bootloader if needed
+#########################################################
+write_bootloader() {
+    if [[ "$arch" == "s390x" ]]; then
+        log "Updating zipl"
+        mkdir -p /mnt/boot_partition
+        mount_boot_partition /mnt/boot_partition
+        trap 'umount /mnt/boot_partition; trap - RETURN' RETURN
+
+        blsfile=$(ls /mnt/boot_partition/loader/entries/*.conf)
+        echo "$(grep options $blsfile | cut -d' ' -f2-) $(cat /tmp/s390x_opts) ignition.firstboot rd.neednet=1 ip=dhcp" > /tmp/zipl_prm
+        zipl --verbose -p /tmp/zipl_prm -i $(ls /mnt/boot_partition/ostree/*/*vmlinuz*) -r $(ls /mnt/boot_partition/ostree/*/*initramfs*) --target /mnt/boot_partition
+        if [[ $? -ne 0 ]]; then
+            log "failed updating bootloder"
+            exit 1
+        fi
+        # this allows a reboot into new installed disk
+        chreipl "${DEST_DEV}"
+    fi
+}
+#########################################################
+#Handle s390x zFCP SCSI devices
+#########################################################
+write_image_to_zfcp_disk() {
+    log "Extracting disk image"
+    mount_tmpfs $(( $($DECOMPRESSION_TOOL ${imagefile_compressed} | wc --bytes) / 1024 / 1024 + ${TMPFS_MBSIZE} ))
+    $DECOMPRESSION_TOOL ${imagefile_compressed} > ${imagefile}
+
+    (while IFS='' read -r line || [ -n "$line" ]
+    do
+        if [[ $line =~ "start=" ]]
+        then
+            # we save the start= and size= here, then dd the partitions later
+            printf "%s:%s\n" \
+                "$(echo $line | cut -d: -f 2 | cut -d, -f 1 | tr -d ' ')" \
+                "$(echo $line | cut -d: -f 2 | cut -d, -f 2 | tr -d ' ')" \
+                >> /tmp/sfdisk_info
+        elif ! [[ $line =~ "label:" || $line =~ "label-id:" || $line =~ "unit:" ]]
+        then
+            continue
+        fi
+        echo $line
+    done <<< $(sfdisk -d ${imagefile})
+    ) | sfdisk "${DEST_DEV}"
+
+    while IFS='' read -r line || [ -n "$line" ]
+    do
+        eval ${line%%:*} ${line##*:}
+        dd if=${imagefile} bs=512 iflag=fullblock \
+            of="${DEST_DEV}" status=progress skip=$start seek=$start count=$size
+        if [[ $? -ne 0 ]]; then
+            log "failed to write image to zFCP SCSI disk"
+            exit 1
+        fi
+    done < /tmp/sfdisk_info
+}
```

Comment:
I tried briefly with `sgdisk -R` which copies the whole disk image including GPT partition table, but was not very sure if it's a sane thing to do. Needs more checking (with s390x team about zFCP SCSI device layout, sgdisk code, and compare data on disk vs image after running `sgdisk -R` using hex dump tool) if this is desired. Advantage is we don't have to extract the image.